### PR TITLE
[TC-184] Org tenancy - delivery services

### PR DESCRIPTION
**Pull request by nir-sopher**
This PR adds the tenancy to the delivery services, and adjusts the API.
No tenancy based authorization yet.
Note that the PR can be pulled on it own, bu merging with PR 398, 402 will improve the test coverage.

### CODE REVIEW

**Comment by nir-sopher** [comment]
All required merges were done.

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst [hunk 1]
```diff
@@ -168,6 +168,8 @@ Delivery Service
   +--------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------+
   | ``sslKeyVersion``        |  int   |                                                                                                                                      |
   +--------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------+
+  | ``tenantId``             | int    | Owning tenant ID                                                                                                                     |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: can you add tenantName as well? and if no tenant, leave tenantName as undef

File: traffic_ops/app/db/migrations/20170404000000_delivery_service_tenancy.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (mitchell852) [inline traffic_ops/app/db/migrations/20170404000000_delivery_service_tenancy.sql]: please change filename to later date

File: traffic_ops/app/lib/Schema/Result/Deliveryservice.pm [hunk 3]
```diff
@@ -278,11 +278,22 @@ __PACKAGE__->table("deliveryservice");
   default_value: false
   is_nullable: 1
 
+=head2 multi_site_origin_algorithm
```
Comment (mitchell852) [inline traffic_ops/app/lib/Schema/Result/Deliveryservice.pm]: not sure why this is in here as it is unrelated to the PR.

File: traffic_ops/app/lib/Schema/Result/Deliveryservice.pm [hunk 4]
```diff
@@ -390,8 +401,12 @@ __PACKAGE__->add_columns(
   { data_type => "text", is_nullable => 1 },
   "logs_enabled",
   { data_type => "boolean", default_value => \"false", is_nullable => 1 },
+  "multi_site_origin_algorithm",
```
Comment (mitchell852) [inline traffic_ops/app/lib/Schema/Result/Deliveryservice.pm]: take this out

File: traffic_ops/app/t/api/1.2/config_files_apache_traffic_server.t [hunk 5]
```diff
@@ -72,7 +72,7 @@ my $count_response = sub {
 	return $t->success( is( scalar(@$r), $count ) );
 };
 
-$t->get_ok('/api/1.2/deliveryservices/list?logsEnabled=true')->status_is(200)->$count_response(2)
+$t->get_ok('/api/1.2/deliveryservices/list?logsEnabled=true')->status_is(200)->$count_response(3)
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/config_files_apache_traffic_server.t]: this doesn't seem related to your PR. you might want to rebase or merge from master and see if this change is still necessary.

File: traffic_ops/app/t/api/1.2/deliveryservice.t [hunk 6]
```diff
@@ -30,302 +30,407 @@ use Test::TestHelper;
...
-		->json_is( "/response/matchList/1/setNumber" => "1")
-		->json_is( "/response/matchList/1/pattern" => ".*\\.my_vod1\\..*")
-	, 'Does the deliveryservice details return?';
-
-# It does not allow changing a regex to be the same as an existing one in another DS in the same CDN
-$ds_id = &get_ds_id('ds_1');
-ok $t->put_ok('/api/1.2/deliveryservices/' . $ds_id . '/update'  => {Accept => 'application/json'} => json => {
-			"xmlId" => "ds_1",
-			"displayName" => "ds_displayname_2",
+sub run_ut {
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/deliveryservice.t]: Is there a reason you put everything inside run_ut(). I don't think it is done this way in other places. And by doing this, I can't really tell what you've changed in this file as it looks like you changed everything.

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst (hunk 1 above)
Comment (nir-sopher) [inline docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: Yes. I plan to do so in a separate PR after Ori's PR for adding it to the user is in.

File: traffic_ops/app/db/migrations/20170404000000_delivery_service_tenancy.sql (hunk 2 above)
Comment (nir-sopher) [inline traffic_ops/app/db/migrations/20170404000000_delivery_service_tenancy.sql]: OK

File: traffic_ops/app/lib/Schema/Result/Deliveryservice.pm (hunk 3 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/Schema/Result/Deliveryservice.pm]: Indeed.
For some reason it was not added in the commit creating multi site origin. 
I tried to close this gap in PR 379 (https://github.com/apache/incubator-trafficcontrol/pull/379).

Would you like me to manually remove the changes from the file myself?

File: traffic_ops/app/t/api/1.2/config_files_apache_traffic_server.t (hunk 5 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/config_files_apache_traffic_server.t]: Actualy it is related. I added a DS with the tenant, and the DS has "logsEnabled". So the count went fom 2 to 3

File: traffic_ops/app/lib/Schema/Result/Deliveryservice.pm (hunk 4 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/Schema/Result/Deliveryservice.pm]: It is auto generated. Wouldn't it be better to except it or pull pr 379 (https://github.com/apache/incubator-trafficcontrol/pull/379) and merge?

File: traffic_ops/app/t/api/1.2/deliveryservice.t (hunk 6 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/deliveryservice.t]: The "run_ut" function was added in order to run the same tests twice. Once with tenancy and once without.
If you want to review it more easily, you can just add "?w=1" at the end of the URL, to review with ignore white-space :
https://github.com/apache/incubator-trafficcontrol/pull/403/files?w=1

**Comment by nir-sopher** [comment]
File was renamed as required

**Comment by nir-sopher** [comment]
@mitchell852 Please note that after rebasing the branch deliveryservice UT fail.
For some reason the changes I do does not work well with the changes you did, moving from &UI::DeliveryService::get_regexp_set to $self->get_regexp_set.
Any clue why?
When I'm making the below change the UT passes


@@ -69,7 +69,7 @@ sub index {
                # build example urls for each delivery service
                my @example_urls = ();
                my $cdn_domain   = $row->cdn->domain_name;
-               my $regexp_set   = $self->get_regexp_set( $row->deliveryservice_regexes );
+               my $regexp_set   = &UI::DeliveryService::get_regexp_set( $self, $row->id );
                @example_urls = &UI::DeliveryService::get_example_urls( $self, $row->id, $regexp_set, $row, $cdn_domain, $row->protocol );
 
                push(
@@ -172,7 +172,7 @@ sub show {
                # build example urls for the delivery service
                my @example_urls = ();
                my $cdn_domain   = $row->cdn->domain_name;
-               my $regexp_set   = $self->get_regexp_set( $row->deliveryservice_regexes );
+               my $regexp_set   = &UI::DeliveryService::get_regexp_set( $self, $row->id );
                @example_urls = &UI::DeliveryService::get_example_urls( $self, $row->id, $regexp_set, $row, $cdn_domain, $row->protocol );

**Comment by nir-sopher** [comment]
Rebase was done.
The below tests are failing regardless to the PR
t/api/1.1/deliveryservice/ssl_keys.t t/user.t t/deliveryserviceserver.t t/api/1.2/deliveryservice_server.t

File: docs/source/development/traffic_ops_api/v12/deliveryservice.rst (hunk 1 above)
Comment (nir-sopher) [inline docs/source/development/traffic_ops_api/v12/deliveryservice.rst]: Ready in another PR
